### PR TITLE
Add Automation Script and Fix for Virtual Inheritance Benchmark Compilation

### PR DESCRIPTION
**Pull request by ViNN280801**
This pull request addresses several improvements and fixes for the benchmark project:

1. **Performance Considerations Documentation**:
   - Added notes about potential performance variations depending on power supply and background processes, particularly for users running benchmarks on laptops.

2. **Automation Script for Benchmark Execution**:
   - Introduced a `bash` script located in the `i9-12900H/v.semykin/g++-14/` directory. This script automates the execution of multiple benchmarks and handles output formatting (removing terminal escape sequences) to generate clean CSV files for further analysis.

3. **Fix for `virtual-inherit` Benchmark Compilation**:
   - Added the missing `virtual-inherit/` subdirectory to the root `CMakeLists.txt`.
   - Resolved a compilation error in the `virtual-inherit` benchmark, which was caused by attempting to pass a `Derived*` to a function expecting a `VirtBase*`. The solution involved introducing a new `startup()` function for the `Base` class hierarchy.

---

#### 1. **Performance Considerations**:

I’ve added documentation on how power settings and background processes might impact benchmark performance:

- **Power Supply**: Running benchmarks on a laptop while using battery power can reduce performance by up to 50%. If the laptop is plugged in, the performance improves, and benchmarks complete faster.
  
- **Background Processes**: Running other tasks while executing benchmarks may reduce performance, as the CPU’s time will be shared among the active processes. Running benchmarks on an idle system is recommended for accurate results.

#### 2. **Automation Script**:

To streamline the benchmark execution process, I created a `bash` script that automates running all benchmarks located in the `i9-12900H/v.semykin/g++-14/` directory. The script ensures that users don’t have to enter commands manually, improving both ease of use and reproducibility.

The script also removes unwanted terminal color codes from the output, which might otherwise lead to "dirty" CSV files.

#### 3. **Virtual Inheritance Benchmark Fix**:

The root `CMakeLists.txt` was missing a reference to the `virtual-inherit/` subdirectory. This caused the benchmarks from that folder not to be built properly. I added the necessary entry to ensure that this benchmark is included during the build process.

Additionally, a compilation error in `virtual-inherit/virtinh.cc` was resolved. The error was caused by passing a `Derived*` (which belongs to a different class hierarchy) to a function expecting a `VirtBase*`. The fix involved adding a separate `startup()` function for the `Base` class hierarchy:

```cpp
int __attribute__((noinline)) startup(Base *b) {
    int sum = 0;
    for (int i = 0; i < NBMK; ++i)
        sum += b->foo(NCALL);
    return sum;
}
```

Now, the virtual-inherit benchmarks compile and execute correctly.

### How to Test:

1. **Run the automation script** to execute all benchmarks and generate CSV output files:

    ```bash
    cd i9-12900H/v.semykin/g++-14/
    ./run_benchmarks.sh
    ```

2. **Review the benchmark results** to ensure that all benchmarks execute successfully and produce the expected output in the CSV files.

3. **Check the `virtual-inherit` benchmarks** to verify that they compile and run without errors.

### CODE REVIEW

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh
Comment (sin-diesel) [inline results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh:1]: Out of curiosity, why did you choose bash as a programmning language for automating benchmarks launch?

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    ["exc_qsort"]="noexcept-qsort/exc_qsort"
+    ["filter"]="ranges-filter/filter"
+    ["projector"]="ranges-projector/projector"
+    ["virtinh"]="virtual-inherit/virtinh"
+    ["virtinl"]="virtual-inline/virtinl"
+    ["virtual"]="virtual-overhead/virtual"
+    ["virtual-shuffle"]="virtual-overhead/virtual-shuffle"
+)
+
+base_dir="../../../../build/Release"
```
Comment (sin-diesel) [inline results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: What happens if I run it from directory other than the one this scripts exists in?

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+for benchmark in "${!benchmarks[@]}"; do
```
Comment (sin-diesel) [inline results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: What if I need to run a set of benchmarks, not all of them? Would be nice to add posibility to do so, better yet, benchmarks need to be labeled. Again, how about we reconsider writing this in Python, for instance?

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh
Comment (ViNN280801) [reply results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh:1]: Just my own convenience. It's pretty simple to write such kind of script and use it on linux

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh (hunk 1 above)
Comment (ViNN280801) [reply results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: > What happens if I run it from directory other than the one this scripts exists in?

Existence error of the executable. Therefore, u need to correct paths within script. It just a sample to help users

File: results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh (hunk 2 above)
Comment (ViNN280801) [reply results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: > What if I need to run a set of benchmarks, not all of them? Would be nice to add posibility to do so, better yet, benchmarks need to be labeled. Again, how about we reconsider writing this in Python, for instance?

Yes, I thought about it and figured out that, for example, we can use flags:

--coro-fibs
--virtinh
etc.

What about Python, yeah, this is a good alternative and more cross-platform than Bash scripting :p
You or we can improve this script to make it more general and convenient in use

Comment (sin-diesel) [reply results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: bash scripts tend to go unmanagable when you add more and more functionality. They are good when  you need to manipulate output of some linux commands, but they barely support some logic beyond that. I would advise rewriting this in python, you could easily label benchmarks and add CLI options for launching them like in any good test/benchmark framework. The issue with paths, CLI and labeling can be much easier solved with python.

Comment (ViNN280801) [reply results/i9-12900H/v.semykin/g++-14/run_benchmarks.sh]: Okay, thank you for this advice, I'll rewrite it later

**Comment by ViNN280801** [comment]
# Benchmark Runner Script Instructions

This Python script automates the process of running benchmarks from executables and managing the generated CSV files. It supports filtering, quiet mode, HDR histograms, and copying CSV files to a specified output directory.

## Usage

```bash
python run_benchmarks.py [OPTIONS]
```

## Options

|       Flag       |                                 Description                                  |
| :-------------- | :-------------------------------------------------------------------------- |
|   -f, --filter   |          Filter benchmarks by the given regular expression pattern.          |
|    -l, --list    |                        List all available benchmarks.                        |
|   -q, --quiet    |              Run benchmarks in quiet mode (no progress output).              |
| -r, --histograms |   Create High Dynamic Range (HDR) Histogram files with a given resolution.   |
| -o, --output-dir | Specify the directory to copy generated CSVs after benchmarks are completed. |
|    --version     |                 Show the program’s version number and exit.                  |

## Benchmark Flags

You can run specific benchmarks by using their flags. Available benchmarks are:

- `--coro_fibs`
- `--excret`
- `--inline`
- `--noexcept_qsort`
- `--ranges_filter`
- `--ranges_projector`
- `--virtual_inherit`
- `--virtual_inline`
- `--virtual_overhead`

## Examples

### Run All Benchmarks

```bash
python run_benchmarks.py
```

### Run Specific Benchmark

Run the `coro_fibs` benchmark:

```bash
python run_benchmarks.py --coro_fibs
```

### Run Specific Benchmark with Filter

Run the coro_fibs benchmark and filter the results:

```bash
python run_benchmarks.py --coro_fibs -f "pattern"
```

### Run Benchmarks in Quiet Mode

Run the coro_fibs benchmark in quiet mode:

```bash
python run_benchmarks.py --coro_fibs -q
```

### Run All Benchmarks and Copy CSVs to a Directory

Run all benchmarks and copy the generated CSVs to the specified directory:

```bash
python run_benchmarks.py -o results/i9-12900H/v.semykin/g++-14/
```

### Run Specific Benchmark and Generate HDR Histograms

Run the virtual_inherit benchmark and generate HDR histograms:

```bash
python run_benchmarks.py --virtual_inherit -r 100
```

## Screenshot examples

### Example 1

```bash
python run_benchmarks.py --virtual_inline -o results/i9-12900H/test
```

image
image
image
image

### Example 2

```bash
python run_benchmarks.py --coro_fibs -r 100 -o results/i9-12900H/test
```
image

## Singal Handling

Script supports handling of the interruption (**Ctrl+C**):

```bash
python run_benchmarks.py --coro_fibs -r 100 -o results/i9-12900H/test
```

```
2024-09-11 19:18:51,403 - INFO - Starting benchmark: coro_fibs at build/Release/coro-fibs/coro-fibs
^C2024-09-11 19:18:53,176 - INFO - Execution interrupted by user. Gracefully stopping the benchmarking process...
2024-09-11 19:18:53,177 - INFO - Removing empty CSV file: ./coro_fibs.csv
```

After interruption it deletes files with size equals 0 bytes.

File: run_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
+import os
```
Comment (sin-diesel) [inline run_benchmarks.py]: Add the default interpreter please

File: run_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    selected_benchmarks = [
+        benchmark for benchmark in benchmarks if getattr(args, benchmark)
+    ]
+
+    if not selected_benchmarks:
+        logging.info(
+            "No specific benchmarks selected. Running all benchmarks recursively from build/Release/"
+        )
+        run_all_benchmarks(
```
Comment (sin-diesel) [inline run_benchmarks.py]: `run_all_benchmarks` should not be a different function, it is same as `run_benchmarks`, the only difference is that `run_all_benchmarks` do not filter anything, that is, regexp is `.*`

File: run_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    "virtual_inherit": ["virtual-inherit/virtinh"],
+    "virtual_inline": ["virtual-inline/virtinl"],
+    "virtual_overhead": [
+        "virtual-overhead/virtual",
+        "virtual-overhead/virtual-shuffle",
+    ],
+}
+
+
+def clean_ansi_escape_sequences(text):
```
Comment (sin-diesel) [inline run_benchmarks.py]: Please use some more object oriented approach: I suggest you create something like smth like `BenchRunner` class and move methods there

File: run_benchmarks.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        text (str): The text with potential ANSI escape sequences.
+
+    Returns:
+        str: Cleaned text without ANSI escape sequences.
+    """
+    ansi_escape = re.compile(r"\x1B[@-_][0-?]*[ -/]*[@-~]")
+    return ansi_escape.sub("", text)
+
+
+def help_msg():
```
Comment (sin-diesel) [inline run_benchmarks.py]: Please make use of `argparse`s native help message instead of writing it yourself

**Comment by sin-diesel** [comment]
Also when you are done with writing your own class please add some basic testing for it, do not forget that benchmarks may not only be run from CLI but also from code, meaning that your class should provide very flexible and definitive API.

**Comment by ViNN280801** [comment]
@sin-diesel, ok, thx, i'll try it later

**Comment by ViNN280801** [comment]
also, what about forgotten virtual inheritance benchs in cmakelists and corrected code in .cc?

**Comment by sin-diesel** [comment]
> also, what about forgotten virtual inheritance benchs in cmakelists and corrected code in .cc?

@tilir

**Comment by ViNN280801** [comment]
In my point of view, i just wanted to write pretty simple script without any OOP-styled code at all, but if it really needs i'll do it with great pleasure. And i didn't want to deal with OOP because it need to create file hierarchy, not just all the code in one place (file), therefore i preferred this functional approach

**Comment by sin-diesel** [comment]
> In my point of view, i just wanted to write pretty simple script without any OOP-styled code at all, but if it really needs i'll do it with great pleasure. And i didn't want to deal with OOP because it need to create file hierarchy, not just all the code in one place (file), therefore i preferred this functional approach

Simple script will eventually die out and will require rewriting it. Using OOP approach and providing a simple but solid architectyre for bench runner framework will ensure this will be supported in the future. Also you do not need to create python project structure here - just one file is enough for now

**Comment by ViNN280801** [comment]
ok, got it

File: run_benchmarks.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+import pandas as pd
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s - %(levelname)s - %(message)s",
+)
+
+
+class BenchRunner:
+    def __init__(self, base_dir="build/Release"):
```
Comment (sin-diesel) [inline run_benchmarks.py:28]: Better to use `pathlib` everywhere when working with paths in python, instead of `str`

**Review by sin-diesel — APPROVED** [review]
There is a lot more to be improved: like type hints, some class hirerarchy changes, defining more precise and flexible interface, some basic CI that runs added tests (thanks for that, by the way), lremoving some code duplication, etc. But generally this looks very nice, LGTM, I suppose we could merge this and do some adjustments later, considering the base code is functional enough, as we do not have the goal of making this launcher perfect in this MR.

**Comment by sin-diesel** [comment]
Also one more important thing: please specify `python` version you test your script with, and please pin the dependencies to `requirements.txt` file (you may also use `poetry` on some other dependency management tool). We need this to be reproducible, and this requires pinning all dependencies.

**Comment by ViNN280801** [comment]
Ok, thx for all the advices, do I need to complete things with requirements in this PR or in further?

**Comment by sin-diesel** [comment]
> Ok, thx for all the advices, do I need to complete things with requirements in this PR or in further? 
> 

Please do it in this MR :)

**Comment by ViNN280801** [comment]
ok

**Review by sin-diesel — APPROVED** [review]
Thanks, nice job adding poetry! Now  we have a pinned  stable environment.

**Comment by tilir** [comment]
@ViNN280801 thanks a lot for your contribution!

**Comment by ViNN280801** [comment]
You are welcome, I always ready and glad to do something useful

**Pull request merged**